src/osp/UserInputHandler.cpp

```diff
@@ -455,4 +466,113 @@ void UserInputHandler::scroll_delta(Vector2i offset)
+    //Load the config
+    for (auto& tab : config.as_table())
+    {
+        std::string name = tab.first;
+        std::vector<ButtonVarConfig> controls;
+        
+        std::vector<ButtonVarConfig> primary = parse_config_string(tab.second.at("primary").as_string().str);
+        controls.insert(controls.end(), primary.begin(), primary.end());
+
+        std::vector<ButtonVarConfig> secondary = parse_config_string((tab.second.at("secondary").as_string().str));
+        controls.insert(controls.end(), primary.begin(), primary.end());
+
+        //Will have to modify the holdable param so that it works
+        config_register_control(name, true,
+            controls);
+    }
+}
+
+std::vector<ButtonVarConfig> UserInputHandler::parse_config_string(std::string_view config)
+{
+    std::vector<ButtonVarConfig> conf;  
+    std::string string = std::string(config);
+    std::stringstream ss(string);
+    std::string token;
+    while (std::getline(ss, token, '+')) {
+        //Parse token
+        ButtonVarConfig button = get_button(token);
+        conf.push_back(button);
+    }
+
+    int size = conf.size() - 1;
+    conf[size].m_nextOp = ButtonVarConfig::VarOperator::OR;
+    conf[size].m_trigger = ButtonVarConfig::VarTrigger::PRESSED;
+
+    return conf;
+}
+
+ButtonVarConfig UserInputHandler::get_button(std::string_view text)
+{
+    using namespace osp;
+    using Key = OSPMagnum::KeyEvent::Key;
+    using Mouse = OSPMagnum::MouseEvent::Button;
+    using VarOp = ButtonVarConfig::VarOperator;
+    using VarTrig = ButtonVarConfig::VarTrigger;
+    ButtonVarConfig config(0, 0, ButtonVar::VarTrigger::HOLD, false, ButtonVar::VarOperator::AND);
+    std::cout << text << std::endl;
+    if (text == "RShift") {
+        config.m_devEnum = (int)Key::RightShift;
+    }
+    else if (text == "LShift") {
+        config.m_devEnum = (int)Key::LeftShift;
+    }
+    else if (text == "LCtrl") {
+        config.m_devEnum = (int)Key::LeftCtrl;
+    }
+    else if (text == "RCtrl") {
+        config.m_devEnum = (int)Key::RightCtrl;
+    }
+    else if (text == "LAlt") {
+        config.m_devEnum = (int)Key::LeftAlt;
+    }
+    else if (text == "RAlt") {
+        config.m_devEnum = (int)Key::RightAlt;
+    }
+    else if (text == "Up") {
+        config.m_devEnum = (int)Key::Up;
+    }
+    else if (text == "Down") {
+        config.m_devEnum = (int)Key::Down;
+    }
+    else if (text == "Left") {
+        config.m_devEnum = (int)Key::Left;
+    }
+    else if (text == "Right") {
+        config.m_devEnum = (int)Key::Right;
+    }
+    else if (text == "RMouse") {
+        config.m_devEnum = (int) Mouse::Right;
+        config.m_device = osp::sc_mouse;
+    }
+    else if (text == "LMouse") {
+        config.m_devEnum = (int)Mouse::Left;
+        config.m_device = osp::sc_mouse;
+    }
+    else if (text == "MMouse") {
+        config.m_devEnum = (int)Mouse::Middle;
+        config.m_device = osp::sc_mouse;
+    }
```

Comment:
Switch statement might be more appropriate here?

Reply:
Unfortunately, C++ doesn't support switch-case for string values.

But what you COULD do, is this.
```
enum class EButton : std::uint8_t
{
    RShift,
    LShift,
    RCtrl,
    LCtrl,
    ....
    Unknown
};
std::string_view button_to_string(EButton button)
{
    switch(button)
    {
        case EButton::RShift: return "RShift";
        case EButton::LShift: return "LShift";
        case EButton::RCtrl: return "RCtrl";
        ....
        default: return "Unknown";
    }
}
EButton string_to_button(std::string_view button)
{
    if(button == "RShift") return EButton::RShift;
    if(button == "LShift") return EButton::LShift;
    if(button == "RCtrl") return EButton::RCtrl;
    return EButton::Unknown;
}
```

Reply:
There's also various "Magic Enum" libraries out there that might be worth taking a look at.

Reply:
I could make a function that converts it back and forth, and I don't think we need an extra enum, because `OSPMagnum::KeyEvent::Key` already provides what we want, we just have to convert it into an int.

Reply:
A map of [string] to structs of {device, enum} is the better to use here. Also good to use that name `string_to_button` as `get_button` sounds like a getter. Something like this should work.

```cpp
struct DeviceButton
{
    int deviceId, int deviceEnum;
}
// ...

// pretend this is a class member
std::map<std::string, DeviceButton> m_strToButton = {
    {"RMouse", {osp::sc_mouse, Mouse::Right}}
    {"LMouse", {osp::sc_mouse, Mouse::Left}}
}

ButtonVarConfig string_to_button(std::string_view buttonName)
{
    // ... something

    auto foundButton = m_strToButton.find(buttonName);

    if (/* device button not found */)
    {
        // do some error thing
    }

    DeviceButton &button = foundButton->second;

    return ButtonVarConfig(button.deviceId, button.deviceEnum, ButtonVar::VarTrigger::HOLD,
                           false, ButtonVar::VarOperator::AND);

}

```

Also consider how other kinds of input devices, like game controllers, can be configured some time in the future. A map would also allow for new keywords can also be added at runtime.

For example, a configuration system dealing with a PS4 controller can do something like this:
1. Read **settings.toml**, and see that the user has configured Controller1 as a PS4 Controller on [OS-specific identifier]
2. Tell some other system to listen to this device in-game, and feed event_raw calls to UserInputHandler
3. Read PS4Controller configs, This contains the device enums
4. Add new keywords to the UserInputHandler Loader's map, such as "Controller1-DPadDown"
5. Read the controls, which can now support stuff like:
```toml
[vehicle_pitch_up]
primary = "S"
secondary = "Controller1-DPadDown"
```